### DISCUSSION TITLE
Is it possible to create a standalone binary?

### DISCUSSION THREAD

**Discussion by stevepryde** [tauri-apps/tauri#3048] [Q&A] [answered]
The build/bundler seems to want to build packages that need to be installed. However I would much prefer a portable app that can be packaged as a single binary and run from the directory without needing to be installed. I'm on Windows 11 FWIW but the question is really more general since it should be possible on any OS.

Is there a way to do this?

**Comment by stevepryde** [comment, marked as answer]
So I discovered I can just run the .exe from src-tauri/target/release and that works, at least in my case.

**Reply by FabianLars** [reply]
Keep in mind that distributing the exe directly is probably not a good idea. You would have to make sure yourself that the users have WebView2 and vcredist installed.
And obviously some stuff won't work, like the sidecar/resource feature or the updater (which will download the MSI) or iirc the notification dialogs will show a PowerShell icon instead of your app icon.

**Reply by vanowm** [reply]
Have you tried run on different PC? It doesn't work for me with the template sample. I can see for a split of a second a window and it closes. I even tried copy entire `release` folder to another PC, it would not work.

**Reply by FabianLars** [reply]
@vanowm Did you read my message above? Just the binary isn't enough for most users.
Like i mentioned above, on windows you need WebView2 and vcredist (both handled by the msi)
On linux you need too much stuff to list here.

Also take a look at discussion #2924

**Comment by begin-theadventure** [comment]
Hi, I'm not sure if I should open a new discussion/issue, but I would like to be able to create a standalone binary, in my case on Linux, because building AppImage and deb wouldn't make sense on the Arch User Repository/AUR or even just for oneself if they want to use just the binary.
I think it would be great to have arguments to specify what to build, something like this:
npm run tauri build -binary / -appimage / -deb

**Reply by FabianLars** [reply]
The binary will always be build as its required for the deb and appimage too, it will be in `target/release/appname`

**Reply by begin-theadventure** [reply]
Never mind, I found this on the AUR :)
tauri.conf.json
```
    "bundle": {
      "active": true,
```
to false

**Comment by vsTianhao** [comment]
Anyway，Making Portable programs is a very important requirement

**Reply by vildantursic** [reply]
Totally agree, I also have requirements from clients to have portable program, without installers.

**Comment by JohnScience** [comment]
I've made some progress in this direction with <https://github.com/JohnScience/tauriless/>.

However, I'm not quite sure how to detect presence of `vcredist`. Also, follow up question, where is the dependency on `vcredist` introduced?

Would it be possible/sensible to use `+crt-static` and `msvc` target on Windows?

**Reply by helgoboss** [reply]
Interested in this as well. I'm writing a sort of installer in Tauri. Needing an installer in order to install an installer is a bit annoying. It would be okay to target Windows 10+.

**Reply by JohnScience** [reply]
By the way, I learned that Rust itself depends on `vcredist` on Windows, unless `+crt-static` is used.

**Comment by gknapp** [comment]
I tried to run my binary on a Windows 11 machine (that ships with Webview2) and has vcredist installed, the behavior is as others describe. What else are the MSI and NSIS installers putting it in place that allow Tauri apps to run?

**Reply by FabianLars** [reply]
> the behavior is as others describe. 

And that is? I don't think i see what you mean in this thread.

A plain tauri app shouldn't require anything. Even vcredist is statically linked nowadays. Are you using any additional rust crates? If you start the app binary from inside a terminal, do you get any errors?

**Reply by helgoboss** [reply]
> > the behavior is as others describe. 
> 
> And that is? I don't think i see what you mean in this thread.
> 
> A plain tauri app shouldn't require anything. Even vcredist is statically linked nowadays. Are you using any additional rust crates? If you start the app binary from inside a terminal, do you get any errors?

I can confirm this. I'm distributing a Tauri 2 app with statically linked runtime on Windows and all it requires is WebView2. Tested with a clean windows installation. Even got it to work on Windows 7 (including all updates) with nothing else than a WebView2 install (though that required building with an older Rust version and downgrading a dependency).

**Reply by gknapp** [reply]
I think this was a false alarm, my app has a dependency on an external resource, that was terminating the app early.

**Comment by ExoticGamerrrYT** [comment]
**Hi! Young developer here.**

I've been researching what we can do to get standalone Tauri executables, and I wanted to share my thoughts here.

The executable generated in `target/release` will work **only** if your program doesn't require any dynamically linked `.dll` files **and** if the machine has [WebView2](https://developer.microsoft.com/en-us/Microsoft-edge/webview2#download) and the [Visual C++ Redistributable](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170#latest-microsoft-visual-c-redistributable-version) installed.

If the app does need `.dll` files, you can still bundle them alongside the `.exe` in a `.zip` archive. But what about WebView and the VC++ redistributable? Well, those are very common dependencies and most machines already have them installed. However, in case they aren’t present, you can still make the app *semi-portable*. Let me explain:

As far as I know, it's not possible to statically link the VC++ runtime, or at least I couldn't get it to work by adding the following to `src-tauri/.cargo/config.toml`. It still ends up dynamically linking the CRT libraries:

```toml
[target.x86_64-pc-windows-msvc]
rustflags = ["-C", "target-feature=+crt-static"]
```

And shipping the entire WebView2 runtime alongside the `.exe` isn’t practical either, since the fixed version folder is over **500MB** in size — not ideal for portability.

So, the only way to distribute a truly *portable* or *standalone* Tauri app on Windows is by:

1. Using the `.exe` from `target/release` if no `.dll` files are required.
2. If your app needs `.dll` files, creating a `.zip` file containing the `.exe` and all required `.dll`s (this can be automated with a simple script), and informing the user of the required dependencies.

That means we can offer users two distribution options:

* A user-friendly installer (MSI or NSIS, generated by Tauri).
* A "portable" `.exe` or `.zip`, as explained above.

We should keep in mind that **we may want to notify users** that WebView2 and the VC++ redistributable must be installed when choosing the standalone `.exe` or `.zip` option.

If anyone finds a way to statically link the CRT or bundle WebView2 into a single `.exe`, I’m all ears! (Or feel free to share any thoughts 😄)
